cadc-data-ops-fits/src/main/java/org/opencadc/fits/slice/NDimensionalSlicer.java

```diff
@@ -0,0 +1,399 @@
+            throws FitsException, IOException {
+        slice(new RandomAccessFileExt(fitsFile, "r"), slices, outputStream);
+    }
+
+    /**
+     * Perform a slice operation from the input RandomAccess.  Implementors will walk through the RandomAccess, skipping
+     * unwanted bytes.
+     *
+     * This implementation relies on the NOM TAM FITS API to be able to look up an HDU by its Extension Name
+     * (<code>EXTNAME</code>) and, optionally, it's Extension Version (<code>EXTVER</code>).  This may affect an
+     * underlying <code>Fits(InputStream)</code> unless the <code>InputStream</code> can handle resetting and marking.
+     *
+     * It is NOT the responsibility of this method to manage the given <code>OutputStream</code>.  The caller will
+     * need to close it and ensure it's open outside the bounds of this method.
+     *
+     * @param randomAccessDataObject         The RandomAccess object to read bytes from.  This method will not close
+     *                                       this file.
+     * @param slices       The string value of the pixels to extract.
+     * @param outputStream Where to write bytes to.  This method will not close this stream.
+     * @throws FitsException    Any FITS related errors from the NOM TAM Fits library.
+     * @throws IOException      Reading Writing errors.
+     */
+    public void slice(final RandomAccessDataObject randomAccessDataObject, final Slices slices,
+                      final OutputStream outputStream)
+            throws FitsException, IOException {
+        final ArrayDataOutput output = new BufferedDataOutputStream(outputStream);
+        final Slices.ExtensionSliceValue[] extensionSliceValues = slices.getExtensionSliceValues();
+
+        LOGGER.debug("Parsed extension slice values: " + Arrays.toString(extensionSliceValues));
+        final Fits fits;
+
+        try {
+            fits = new Fits(randomAccessDataObject);
+        } catch (FitsException fitsException) {
+            throw new IllegalStateException(fitsException.getMessage(), fitsException);
+        }
+
+        slice(fits, extensionSliceValues, output);
+    }
+
+    private void slice(final Fits fits, final Slices.ExtensionSliceValue[] extensionSliceValues,
+                       final ArrayDataOutput output) throws FitsException, IOException {
+        // Assume MEF if more than one extension was requested, so write a Primary HDU before anything else.
+        if (extensionSliceValues.length > 1) {
+            final BasicHDU<?> firstHDU = fits.getHDU(0);
+            final Header headerCopy = ImageHDU.manufactureHeader(firstHDU.getData());
+            final Header primaryHeader = firstHDU.getHeader();
+
+            copyHeader(primaryHeader, headerCopy);
+
+            final HeaderCard nextEndCard = headerCopy.findCard(DataDescription.NEXTEND);
+
+            // Adjust the NEXTEND appropriately.
+            if (nextEndCard == null) {
```

Comment:
This is going to be problematic. At this point we do not know how many extensions are going to be output so we do not know what value to set NEXTEND to. For example, the sample-mef.fits has a primary and 3 extensions: what if the caller asks for pixels from extension 7? This code would set NEXTEND = 1 but then not be able to output an extension; the output would be invalid. 

Yeah, fits has several things like this that make streaming output hard and this is one of them. The first thing is to verify what happens when NEXTEND is missing or wrong (both from a spec point of view and whther nom-tam-fits and fitsverify think it is OK or complain).

To be strictly correct, one would have to figure out how many extensions will actually be output, which means iterating through the HDUs until you find the ones that match the requested slices. It looks like the Fits object keeps the already-read HDUs in a list and that `getHDU(n)` re-uses that so you can jump back without re-reading. 

Note: It remains to be verified what IO ops occur when doing this and if going back and slicing data works as expected. It is clear that an MEF file with many extensions will potentially store a lot of header data in memory (something to worry about later). It's good to keep an eye on the IO ops and see when a change in code changes the IO ops being performed.

So, in addition to dealing with the 3 types of FITS files above, I would suggest a loop like the one in `FitsOperations.getHeaders()` that checks each header to determine if it is to be part of the output (and stopping loop as soon as all requested extensions are accounted for). After that point, you will know the right value for NEXTEND, will be able to output extensions in either original (file) order or requested (slice) order, and -- when we do get to applying sky coord cutouts -- we could have already done the intersection of the sky shape with the data array.

Reply:
mistakenly mistyped inherent instead of inherit in the earlier review (fixed now for clarity). Exentions in MEF files of type 2 inherit metadata from the primary (can't recall which has precedence if a header is set in both - I'd assume the most local but that could be wrong because of reasons).